modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessScript.java

```diff
@@ -56,6 +57,9 @@
      * @return The generated ScriptException.
      */
     default ScriptException convertToScriptException(Throwable t, Map<String, List<String>> extraMetadata) {
+        if (t instanceof DroppedDocumentException) {
```

Comment:
Painless shouldn't know about ingest. I think we need to discuss with @jdconrad how we can change painless to allow some exception types to be propagated verbatim instead of wrapping in a ScriptException. Perhaps we could allow `execute` on script classes to have checked exceptions, and those exceptions would be rethrown instead of wrapped?

Reply:
@rjernst yea this was really dirty, hence the `WIP` label.

@jdconrad @rjernst I like the idea of a checked `Exception` for this. Maybe something like `ScriptExitException`? We could even add an exit code field to it and maybe reuse it for other scenarios as well (thinking about ingest processors called from Painless directly failing here mostly ... those currently generate a weirdly complicated wrapped exception when failing, when in actuality you probably just want the failure in the processor to propagate up cleanly)?

Then we could do:

```java
} catch (ScriptExitException e) {
   if (e.exitCode() == 0) {
      // Drop document quietly
   } else {
     // Log error and do other error things
   }
}
```

WDYT?

Reply:
You can't use exceptions in this form because a user could potentially catch them in the script.  We decided a while back that if we want to get around this we can use Error classes and we will never whitelist them.  These bypass the users ability to catch them in script.

Reply:
I don't think an error is right here. This is essentially an exception the user is throwing (through calling the drop() method). Sure, they could catch it themselves. Another option (suggested by @jdconrad privately) is to have a flag on the IngestDocument, and set this, kind of like we have noop for update scripts. This might be an ok solution; we would then make the drop method take the ingest document and set the flag.

To be clear, though, my proposal was to change the execute method to something like this:

```
void execute(Map ctx) throws DroppedDocmentException
```

Painless would determine any checked exceptions when reflecting the execute method, and add a catch/rethrow into the existing try/catch within the main execute implementation. So the exception type would not be generic, but still specific to ingest, and painless would learn about it through reflection (as it does the other argument types as well).